image-hd.c

```diff
@@ -590,9 +602,17 @@ static int hdimage_setup(struct image *image, cfg_t *cfg)
 	}
 
 	if (hd->partition_table) {
-		now = 512;
-		if (hd->gpt)
-			now = hd->gpt_location + (GPT_SECTORS - 1) * 512;
+		struct partition *mbr = fake_partition("[MBR]", 0, 512);
+
+		list_add(&mbr->list, &image->partitions);
+		if (hd->gpt) {
+			struct partition *gpt_header, *gpt_table;
+
+			gpt_header = fake_partition("[GPT header]", 512, 512);
+			gpt_table = fake_partition("[GPT table]", hd->gpt_location, (GPT_SECTORS - 1) * 512);
+			list_add(&gpt_header->list, &mbr->list);
+			list_add(&gpt_table->list, &gpt_header->list);
+		}
```

Comment:
Hmm, I'm not so sure about this one. On the one hand, this looks quite elegant, but on the other hand, we still have basically the same checking with the backup GPT at the end. And adding a fake partition for this is not that simple, right?

Reply:
I think that's doable. If the image has a fixed size, we add a fake partition representing the backup (and add it last in the partition list). That will automatically do the overlap checking once the loop gets to that last fake partition; it will be compared against all the previous ones. If the image doesn't have a fixed size, just as we do now, the image size if automatically extended to include room for the backup, so it really can't overlap anything.